# Perform probe pruning for the duplication-only model

#################################################
### Libraries ###################################
library(data.table)
library(dplyr)
library(tidyr)
library(stringr)


#################################################
### External arguments ##########################

args <- commandArgs(trailingOnly = TRUE)
chr <- args[1]
print(paste0("Analyzing chromosome ", chr))


#################################################
### STEP 1: Load & filter #######################

# Load all genotype frequencies; "genotype_freq_WB_All.txt.gz" is generated by the script "GWAS/02_probes/03_genotype_frequency.R"
geno_freq <- as.data.frame(fread("/home/cauwerx/scratch/cauwerx/projects/cnv_gwas/gwas/02_probes/white_british/stringent/frequency/data/final/genotype_frequency/genotype_freq_WB_All.txt.gz", header = T))

# Filter: Threshold for frequency (duplication): >= 0.005%
geno_freq <- geno_freq[which(geno_freq$Chr == chr & geno_freq$FreqDup >= 0.005), ]
print(paste0("Number of probes with duplication frequency >= 0.005% on chromosome ", chr, ": ", nrow(geno_freq)))

# Save temporarily
fwrite(geno_freq[, 2, drop = F], paste0("/home/cauwerx/scratch/cauwerx/projects/cnv_gwas/gwas/02_probes/white_british/stringent/duplication_only/data/temp/probes_WB_All_dup_0.005_chr", chr, ".txt"), col.names = F, row.names = F, quote = F, sep = "\t")


#################################################
### STEP 2: Pruning with PLINK v2 ###############

# Test different thresholds for r^2: 0.9, 0.99, 0.999, 0.9999
thr_r <- c(0.9, 0.99, 0.999, 0.9999)

# Loop over tested thresholds; 0.9999 was used for further analyses 
for (i in thr_r) {

	# Prepare files
	input <- paste0("/data/FAC/FBM/DBC/zkutalik/default_sensitive/cauwerx/CNV_GWA/duplication_only/ukb_cnv_bed/ukb_cnv_chr", chr) # Duplication PLINK file set for the given chromosome
	probes <- paste0("/home/cauwerx/scratch/cauwerx/projects/cnv_gwas/gwas/02_probes/white_british/stringent/duplication_only/data/temp/probes_WB_All_dup_0.005_chr", chr, ".txt") # Output from the frequency filtering (line 30)
	output <- paste0("/home/cauwerx/scratch/cauwerx/projects/cnv_gwas/gwas/02_probes/white_british/stringent/duplication_only/data/final/pruning/r_", i,"/probes_WB_All_prune_", i,"_dup_chr", chr)

	# Run PLINK v2
	system(paste("/home/cauwerx/scratch/cauwerx/softwares/plink2/plink2",
				 "--bfile", input,
				 "--keep /home/cauwerx/scratch/cauwerx/projects/cnv_gwas/gwas/01_samples/white_british/data/final/samples_white_british_All.txt",
				 "--extract", probes,
				 "--indep-pairwise 500 250", i,
				 "--out", output))

	# Delete the log file
	unlink(paste0(output, ".log"))

	# Assess PLINK run
	print(paste0("Number of remaining probes on chromosome ", chr, " after frequency (duplication-only) selection and pruning at r ", i, ": ", nrow(as.data.frame(fread(paste0(output, ".prune.in"), header = F)))))

}


#################################################
### STEP 3: Delete temporary file ###############

unlink(paste0("/home/cauwerx/scratch/cauwerx/projects/cnv_gwas/gwas/02_probes/white_british/stringent/duplication_only/data/temp/probes_WB_All_dup_0.005_chr", chr, ".txt"))
